t/12_sles4sap_publicccloud.t

```diff
@@ -484,4 +485,57 @@ subtest '[get_hana_site_names] values from settings' => sub {
     ok(($res[1] eq 'ZhengHe'), "Value for the secondary site is from setting");
 };
 
+subtest '[wait_for_zypper] zypper unlocked at first try' => sub {
+    my $self = sles4sap_publiccloud->new();
+    my $pc_instance = Test::MockModule->new('publiccloud::instance');
+    my $instance = publiccloud::instance->new();
+    $pc_instance->redefine(run_ssh_command => sub { return 0; });
+
+    lives_ok { $self->wait_for_zypper($instance) } 'Zypper was not locked, command succeeded without retries';
+};
+
+subtest '[wait_for_zypper] zypper fails at first try with non 7 rc' => sub {
+    my $self = sles4sap_publiccloud->new();
+    my $pc_instance = Test::MockModule->new('publiccloud::instance');
+    my $instance = publiccloud::instance->new();
+    $pc_instance->redefine(run_ssh_command => sub { return 1; });
+
+    lives_ok { $self->wait_for_zypper($instance) } 'Zypper command failed with a non-locking issue and did not retry';
+};
+
+subtest '[wait_for_zypper] zypper fails at first try with 7 rc but pass at second retry' => sub {
+    my $sles4sap_publiccloud = Test::MockModule->new('sles4sap_publiccloud');
+    my $self = sles4sap_publiccloud->new();
+    my $pc_instance = Test::MockModule->new('publiccloud::instance');
+    my $instance = publiccloud::instance->new();
+    my $attempt = 0;
+    my @record_infos;
+
+    $pc_instance->redefine(run_ssh_command => sub {
+            return $attempt++ ? 0 : 7;    # return 7 on first call, 0 on second
+    });
+    $sles4sap_publiccloud->redefine(record_info => sub {
+            push @record_infos, join(' ', @_);
+    });
+
+    lives_ok { $self->wait_for_zypper($instance) } 'Zypper was locked initially but succeeded on retry';
+    ok(any { /Zypper is locked/ } @record_infos, "record_info called with expected text");
+};
+
+subtest '[wait_for_zypper] zypper fails always with 7 rc' => sub {
+    my $self = sles4sap_publiccloud->new();
+    my $sles4sap_publiccloud = Test::MockModule->new('sles4sap_publiccloud');
+    my $pc_instance = Test::MockModule->new('publiccloud::instance');
+    my $instance = publiccloud::instance->new();
+    my @record_infos;
+
+    $pc_instance->redefine(run_ssh_command => sub { return 7; });
+    $sles4sap_publiccloud->redefine(record_info => sub {
+            push @record_infos, join(' ', @_);
+    });
+
+    dies_ok { $self->wait_for_zypper($instance, max_retries => 3) } 'Zypper remained locked after max retries';
+    is(scalar @record_infos, 3, 'wait_for_zypper was retried exactly 3 times');
```

Comment:
my 2 cents: do not do it. It is not behavioral, it is just debug. The `wait_for_zypper` is only using `record_info`  for debug purpose.

Reply:
yes, but also `record_info` is called every time rc 7 is returned. So this is a way to count how many times the command encountered `rc 7` before it aborted, and check that it is the number defined by the user. But I can remove it and only check that in the end, the function died as it's supposed to.